runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/observers/DirectStreamObserver.java

```diff
@@ -89,6 +90,9 @@ public void onNext(T value) throws StreamObserverCancelledException {
             throw new StreamObserverCancelledException("StreamObserver was terminated.");
           }
 
+          // We close under "lock", so this should never happen.
```

Comment:
needs note about terminating phaser first, update below too.

// We close under "lock" after terminating, so if we didn't observer termination above
// we shouldn't be closed.

Reply:
done